helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -490,66 +495,75 @@ private boolean updateOnHostChange(final DeploymentGroup group,
-        curHostsVersion = Optional.of(chn.getStat().getVersion());
-        curHosts = Json.read(chn.getBytes(), new TypeReference<List<String>>() {});
-      } catch (JsonMappingException e) {
-        curHosts = Collections.emptyList();
-      }
-
       final DeploymentGroupStatus status = getDeploymentGroupStatus(groupName);
       if (!allowHostChange(status)) {
         return;
       }
-      if (!curHostsVersion.isPresent() || !hosts.equals(curHosts)) {
-        // Node not present or hosts have changed
-        final List<ZooKeeperOperation> ops = Lists.newArrayList();
-        ops.add(set(Paths.statusDeploymentGroupHosts(groupName), Json.asBytes(hosts)));
-
-        final Node dgn = client.getNode(Paths.configDeploymentGroup(groupName));
-        final Integer deploymentGroupVersion = dgn.getStat().getVersion();
-        DeploymentGroup deploymentGroup = Json.read(dgn.getBytes(), DeploymentGroup.class);
-
-        ImmutableList<Map<String, Object>> events = ImmutableList.of();
-
-        if (deploymentGroup.getJobId() != null) {
-          if (updateOnHostChange(deploymentGroup, status)) {
-            deploymentGroup = deploymentGroup.toBuilder()
-                .setRollingUpdateReason(HOSTS_CHANGED)
-                .build();
-            // Fail transaction if the deployment group has been updated elsewhere.
-            ops.add(check(Paths.configDeploymentGroup(groupName), deploymentGroupVersion));
-
-            // NOTE: If the DG was removed this set() cause the transaction to fail, because
-            // removing the DG removes this node. It's *important* that there's an operation that
-            // causes the transaction to fail if the DG was removed or we'll end up with
-            // inconsistent state.
-            ops.add(set(Paths.configDeploymentGroup(groupName), deploymentGroup));
-
-            final RollingUpdateOp op = getInitRollingUpdateOps(deploymentGroup, hosts, client);
-            ops.addAll(op.operations());
-            events = op.events();
-          }
-        }
 
-        log.info("starting zookeeper transaction for updateDeploymentGroupHosts on "
-                 + "deployment-group name {} jobId={}. List of operations: {}",
-            groupName, deploymentGroup.getJobId(), ops);
+      final List<String> curHosts = getHosts(client, Paths.statusDeploymentGroupHosts(groupName));
+      final List<String> previouslyRemovedHosts = getHosts(
+          client, Paths.statusDeploymentGroupRemovedHosts(groupName));
+
+      if (hosts.equals(curHosts)) {
+        return;
+      }
+
+      final List<String> removedHosts = removedHosts(curHosts, hosts, previouslyRemovedHosts);
+
+      final List<ZooKeeperOperation> ops = Lists.newArrayList();
+      ops.add(set(Paths.statusDeploymentGroupHosts(groupName), Json.asBytes(hosts)));
+      ops.add(set(Paths.statusDeploymentGroupRemovedHosts(groupName), Json.asBytes(removedHosts)));
```

Comment:
Discussed this with @mattnworb in person. There's only one condition I can envision where this list will grow unbounded, which is if an unbounded amount of agents are removed from a DG (i.e. the agents are still running, but with labels that no longer match the DG) *and* we're unable to undeploy the job from those machines for some reason. In this case the failure to undeploy will cause the rolling update to fail, so hopefully the DG owner will notice and take action.

I was also concerned about the case where agents disappear permanently without being degregistered from Helios (i.e. the agents' hosts are decommissioned without deregistering). This *would* cause the list to grow unbounded, except that we have code to automatically deregister these AWOL agents after ~7 days, at which point the undeploy will 'succeed' because the deployment of the DG's job to the agent will have been deleted.

I'm going to add a test for the condition explained in the second paragraph, just to be sure.